compiler-rt/src/crt0/allocator.cairo

```diff
@@ -0,0 +1,413 @@
+                // next allocation by the requested amount and return the current pointer.
+                self.next_allocation += byte_count;
+
+                // If the allocation uses up the remaining space, mark the cell as fully consumed.
+                if byte_count == bytes_left_in_current_cell {
+                    self.space_in_current_cell = false;
+                }
+
+                // No need to allocate a new cell, because we're still in the previous one.
+                return current_allocation;
+            } else {
+                // The allocation request does not fit in the current cell. Make a note of this
+                // sub-cell region, so it's available for the next allocation.
+                let sub_cell_region_size = bytes_left_in_current_cell.into();
+                let sub_cell_region_address = self.next_allocation.into();
+                self
+                    .allocated_sub_cell_regions
+                    .insert(sub_cell_region_size, sub_cell_region_address);
+                self.sub_cell_regions_count += 1;
+
+                // Move the allocation pointer to the beginning of the next cell, from where a
+                // larger region can be allocated.
+                self.next_allocation += bytes_left_in_current_cell;
+                self.space_in_current_cell = false;
+            }
+        }
+
+        // Here, the next allocation pointer must be at the beginning of a new cell.
+        let current_allocation = self.next_allocation;
+        self.next_allocation += byte_count;
+
+        if byte_count % SIZEOF_CELL != 0 {
+            // Make a note we're not fully utilizing the last cell, so we may want to use it for
+            // the next allocation.
+            self.space_in_current_cell = true;
+        }
+
+        // Allocate cells in the mapping
+        let cells_needed = utils::cells_count_from_bytes(current_allocation, byte_count);
+        let mut cells_allocated: u128 = 0;
+        while cells_allocated < cells_needed {
+            let key: felt252 = (current_allocation + cells_allocated * SIZEOF_CELL).into();
+            const DEFAULT_MEMORY_VALUE: felt252 = 0;
+            self.allocated_addresses.insert(key, DEFAULT_MEMORY_VALUE);
+            cells_allocated += 1;
+        };
+
+        current_allocation
+    }
+
+    fn load(ref self: AllocatorState, address: Address, byte_count: ByteCount) -> Array<u8> {
+        let cells_needed = utils::cells_count_from_bytes(address, byte_count);
+        let offset_in_first_cell: Address = address % SIZEOF_CELL;
+        let cell_base_addr = utils::cells_base_address(address);
+        let mut cells_read: u128 = 0;
+        let mut bytes_to_read = byte_count;
+
+        // Buffer to store bytes of retrieved memory cells.
+        let mut output = array![];
+
+        while cells_read < cells_needed && bytes_to_read > 0 {
+            let key = (cell_base_addr + cells_read * SIZEOF_CELL);
```

Comment:
These parentheses seem strange? Might make sense to paren the `(cells_read * SIZEOFF_CELL)` just to be clearer, but I'd remove them if you don't want to do that.